provisioner/terraform/resources.go

```diff
@@ -684,41 +695,69 @@ func ConvertState(ctx context.Context, modules []*tfjson.StateModule, rawGraph s
 			if err != nil {
 				return nil, xerrors.Errorf("decode metadata attributes: %w", err)
 			}
-			resourceLabel := convertAddressToLabel(resource.Address)
 
-			var attachedNode *gographviz.Node
-			for _, node := range graph.Nodes.Lookup {
-				// The node attributes surround the label with quotes.
-				if strings.Trim(node.Attrs["label"], `"`) != resourceLabel {
-					continue
+			var targetLabel string
+
+			// First, check if ResourceID is provided and try to find the resource by ID
+			if attrs.ResourceID != "" {
+				// Look for a resource with matching ID
+				foundLabels := labelsByResourceID[attrs.ResourceID]
+				switch len(foundLabels) {
+				case 0:
+					// If we couldn't find by ID, fall back to graph traversal
+					logger.Warn(ctx, "coder_metadata resource_id not found, falling back to graph traversal",
+						slog.F("resource_id", attrs.ResourceID),
+						slog.F("metadata_address", resource.Address))
+				case 1:
+					// Single match - use it
+					targetLabel = foundLabels[0]
+				default:
+					// Multiple resources with same ID - this creates ambiguity
+					logger.Warn(ctx, "multiple resources found with same resource_id, falling back to graph traversal",
+						slog.F("resource_id", attrs.ResourceID),
+						slog.F("metadata_address", resource.Address),
+						slog.F("matching_labels", foundLabels))
 				}
-				attachedNode = node
-				break
 			}
-			if attachedNode == nil {
-				continue
-			}
-			var attachedResource *graphResource
-			for _, resource := range findResourcesInGraph(graph, tfResourcesByLabel, attachedNode.Name, 0, false) {
-				if attachedResource == nil {
-					// Default to the first resource because we have nothing to compare!
-					attachedResource = resource
-					continue
+
+			// If ResourceID wasn't provided or wasn't found, use graph traversal
+			if targetLabel == "" {
+				resourceLabel := convertAddressToLabel(resource.Address)
+
+				var attachedNode *gographviz.Node
+				for _, node := range graph.Nodes.Lookup {
```

Comment:
At this stage, it would make sense to extract the "metadata association" piece out and test it separately.